batchspawner/batchspawner.py

```diff
@@ -775,4 +777,106 @@ def state_gethost(self):
         self.log.error("Spawner unable to match host addr in job {0} with status {1}".format(self.job_id, self.job_status))
         return
 
+
+class OARSpawner(UserEnvMixin, BatchSpawnerBase):
+    batch_script = Unicode("""#!/bin/sh
+#OAR -n jupyterhub-singleuser-{{username}}
+#OAR -l nodes=1{% if nprocs     %}/core={{nprocs}}{% endif %}{% if runtime    %},walltime={{runtime}}{% endif %}
+{% if queue      %}#OAR -q {{queue}}{% endif %}
+{% if memory     %}#OAR -p "memnode>={{memory}}"{% endif %}
+{% if options    %}#OAR {{options}}{% endif %}
+
+{{prologue}}
+{{cmd}}
+{{epilogue}}
+""").tag(config=True)
+
+    # use a script to submit to batch queueing system instead of piping it in
+    script_mode = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IROTH | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH #755 by default
+
+    script_home = os.path.join(os.environ.get("HOME", "/tmp"), "oar_jobs")
+    if not os.path.exists(script_home):
+        os.makedirs(script_home)
+    # outputs job id string
+    batch_submit_cmd = Unicode('/usr/bin/oarsub -d {homedir} -S ').tag(config=True)
+    # outputs job data XML string
+    batch_query_cmd = Unicode('/usr/bin/oarstat -J -j {job_id}').tag(config=True)
+    batch_cancel_cmd = Unicode('/usr/bin/oardel {job_id}').tag(config=True)
+
+    @gen.coroutine
+    def run_command(self, cmd, input=None, env=None):
+        if input: 
+            # Use script file
+            script = os.path.join(self.script_home, "jupyter_%s.sh"%self.get_req_subvars()["username"])
```

Comment:
Why don't you use the Python facilities to securely make tempfiles?  Making files with predictable names can have security risks.

Does the script file need to exist until the whole job is done, or can it be removed right away?  If it can be removed as soon as the submit command returns, it can be done with a context manager too:

```
def run_command(self, cmd, input=None, env=None):
    with  tempfile.NamedTemporaryFile(dir=self.script_home, prefix="jupyter_%s"%username, suffix=".sh") as scriptfile:
        ... Save the script file
        ... modify `cmd` to include that script file
        return super().run_command(cmd, input=None, env=env)
```

Then it'll be automatically cleaned up, and made securely.  There are more options if needed.

Reply:
Unfortunately, this file is actually re-read when the runner starts ... so it cannot be removed straight away. 
I believe this is a bug within OAR and I found no workaround. Apparently there is little concern on security within OAR.

Reply:
Heh, fair enough.  If files are made in admin's home dir and not /tmp, the predictable file names should be OK.  Let's just make sure it's documented.